tests/test_jwt.py

```diff
@@ -126,6 +126,23 @@ def test_decode_valid(token_factory):
     assert payload["metadata"]["meta"] == "data"
 
 
+def test_decode_header_object(token_factory):
+    payload = token_factory()
+    payload = b"M7." + b".".join(payload.split(b".")[1:])
+
+    with pytest.raises(ValueError) as excinfo:
+        jwt.decode(payload, certs=PUBLIC_CERT_BYTES)
+    assert excinfo.match(r"Header segment should be a JSON object: b'M7'")
+
+
+def test_decode_payload_object(signer):
+    payload = jwt.encode(signer, "iatexp")
```

Comment:
Please add a comment to this as well, and how it correlates to the error message